source/SVInt.cpp

```diff
@@ -237,39 +177,76 @@ SVInt::SVInt(uint16_t bits, LiteralBase base, bool isSigned, bool anyUnknown, Ar
             break;
         case LiteralBase::Decimal:
             radix = 10;
+            shift = 0; // can't actually shift for Decimal numbers
             break;
         case LiteralBase::Hex:
             radix = 16;
             shift = 4;
             break;
-        default:
-            ASSERT(false, "Unreachable");
-            break;
+
+        DEFAULT_UNREACHABLE;
+    }
+
+    if (isSingleWord()) {
+        // Fast path for values that fit in one word:
+        // let's do the computation entirely in a normal uin64_t instead
+        // of having to do SVInt operations
+        val = 0;
+        for (const logic_t& d : digits) {
+            if (shift)
+                val <<= shift;
+            else
+                val *= radix;
+            val += d.value;
+            ASSERT(d.value < radix, "Digit %d too large for radix %d", d.value, radix);
+        }
+        // handles overflow
+        clearUnusedBits();
+        return;
     }
 
+    pVal = new uint64_t[getNumWords()]();
+
     int ones = (1 << shift) - 1;
 
     SVInt digit(bitWidth, 0, false);
     SVInt radixSv(bitWidth, radix, false);
 
+    // If the user specified a number too large to fit in the number of bits specified,
+    // the spec says to truncate from the left, which this method will successfully do.
     for (const logic_t& d : digits) {
         int unknown = 0;
         int value = d.value;
-        if (d == logic_t::x) {
+
+        if (exactlyEqual(d, logic_t::x)) {
             value = 0;
             unknown = ones;
         }
-        else if (d == logic_t::z) {
+        else if (exactlyEqual(d, logic_t::z)) {
             value = ones;
             unknown = ones;
+        } else {
+            ASSERT(value < radix, "Digit %d too large for radix %d", value, radix);
         }
 
         // shift if we can, multiply if we must
+        // this operation can result in us losing our unknown flag, so let's
+        // make sure that doesn't happen
+        bool unknownFlagStore = unknownFlag;
+        int bitWidthStore = bitWidth;
         if (shift)
             *this = shl(shift, false);
         else
             *this *= radixSv;
```

Comment:
I think it's probably worth in the future having an internal mul() method that, like shl(), takes a flag indicating whether the unknown flag should go away. That way we don't have to do this extra checking here.